assembly/src/library/mod.rs

```diff
@@ -15,6 +20,150 @@ pub use self::version::{Version, VersionError};
+impl CompiledLibrary {
+    /// Constructs a new [`CompiledLibrary`].
+    pub fn new(
+        mast_forest: MastForest,
+        exports: Vec<FullyQualifiedProcedureName>,
+    ) -> Result<Self, CompiledLibraryError> {
+        if mast_forest.num_procedures() as usize != exports.len() {
+            return Err(CompiledLibraryError::InvalidExports {
+                exports_len: exports.len(),
+                roots_len: mast_forest.num_procedures() as usize,
+            });
+        }
+
+        Ok(Self {
+            mast_forest,
+            exports,
+        })
+    }
+}
+
+impl CompiledLibrary {
+    /// Returns the inner [`MastForest`].
+    pub fn mast_forest(&self) -> &MastForest {
+        &self.mast_forest
+    }
+
+    /// Returns the fully qualified name of all procedures exported by the library.
+    pub fn exports(&self) -> &[FullyQualifiedProcedureName] {
+        &self.exports
+    }
+
+    /// Returns an iterator over the module infos of the library.
+    pub fn into_module_infos(self) -> impl Iterator<Item = ModuleInfo> {
+        let mut modules_by_path: BTreeMap<LibraryPath, ModuleInfo> = BTreeMap::new();
+
+        for (proc_index, proc_name) in self.exports.into_iter().enumerate() {
+            modules_by_path
+                .entry(proc_name.module.clone())
+                .and_modify(|compiled_module| {
+                    let proc_node_id = self.mast_forest.procedure_roots()[proc_index];
+                    let proc_digest = self.mast_forest[proc_node_id].digest();
+
+                    compiled_module.add_procedure_info(ProcedureInfo {
+                        name: proc_name.name.clone(),
+                        digest: proc_digest,
+                    })
+                })
+                .or_insert_with(|| {
+                    let proc_node_id = self.mast_forest.procedure_roots()[proc_index];
+                    let proc_digest = self.mast_forest[proc_node_id].digest();
+                    let proc = ProcedureInfo {
+                        name: proc_name.name,
+                        digest: proc_digest,
+                    };
+
+                    ModuleInfo::new(proc_name.module, vec![proc])
+                });
+        }
+
+        modules_by_path.into_values()
+    }
+}
+
+// MODULE INFO
+// ===============================================================================================
+
+/// Stores a module's path, as well as information about all exported procedures.
+#[derive(Debug, Clone)]
+pub struct ModuleInfo {
+    path: LibraryPath,
+    procedure_infos: Vec<ProcedureInfo>,
+}
+
+impl ModuleInfo {
+    /// Constructs a new [`ModuleInfo`].
+    pub fn new(path: LibraryPath, procedures: Vec<ProcedureInfo>) -> Self {
+        Self {
+            path,
+            procedure_infos: procedures,
+        }
+    }
+
+    /// Adds a [`ProcedureInfo`] to the module.
+    pub fn add_procedure_info(&mut self, procedure: ProcedureInfo) {
+        self.procedure_infos.push(procedure);
+    }
```

Comment:
Should this be a private method? Or are we planning to use it outside of this module?

Reply:
Turns out `ModuleInfo` was not needed in this PR, so I removed it. I can add it again in the subsequent PR if it is still relevant